core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -256,24 +278,61 @@ class AvroUpsertMonitorActor(
-        }.unsafeToFuture
+  private def initUpsert(upsertFile: String, jobId: UUID, ackId: String, workspaceName: WorkspaceName, userInfo: UserInfo): Future[ImportUpsertResults] = {
+    logger.info(s"beginning upsert process for $jobId ...")
+
+    // Start reading the file. This returns a stream
+    logger.info(s"checking access to $upsertFile for jobId ${jobId.toString} ...")
+    val upsertStream = getUpsertStream(upsertFile)
+
+    // ensure that the file has some contents. Our implementation of GoogleStorageInterpreter will return an empty
+    // stream instead of an error in cases where it could not read the file.
+    if (Stream.empty == upsertStream) {
+      throw new RawlsExceptionWithErrorReport(RawlsErrorReport(StatusCodes.BadRequest, s"Batch upsert file $upsertFile not found."))
+    }
+
+    // Ack the pubsub message as soon as we know we can read the file. pro: don't have to worry about ack timeouts for long operations, con: if someone restarts rawls here the upsert is lost
+    logger.info(s"acking upsert pubsub message for jobId ${jobId.toString} ...")
+    acknowledgeMessage(ackId)
+    // translate the stream of bytes to a stream of json
+    val jsonStream = upsertStream.through(byteArrayParser)
+    // translate the stream of json to a stream of EntityUpdateDefinition. This is awkward, because we break out
+    // of Circe and use spray-json parsing, since we have complex custom spray-json decoders for EntityUpdateDefinition
+    // and I would prefer not to have multiple complex custom decoders
+    val entityUpdateDefinitionStream = jsonStream.map { circeJson =>
+      circeJson.toString().parseJson.convertTo[EntityUpdateDefinition]
+    }
+
+    // chunk the stream into batches.
+    val batchStream = entityUpdateDefinitionStream.chunkN(batchSize)
+
+    // upsert each chunk
+    val upsertFuturesStream = batchStream map { chunk =>
+      val upsertBatch: List[EntityUpdateDefinition] = chunk.toList
+      logger.info(s"inserting a batch of ${upsertBatch.size} for jobId ${jobId.toString} ...")
+      toFutureTry(entityService.apply(userInfo).batchUpdateEntitiesInternal(workspaceName, upsertBatch, true))
+    }
+
+    // wait for all upserts
+    Future.sequence(upsertFuturesStream.compile.toList.unsafeRunSync()) map { upsertResults =>
+
+      val numSuccesses:Int = upsertResults.collect {
+        case Success(entityList) => entityList.size
+      }.sum
+
+      val failureReports:List[RawlsErrorReport] = upsertResults collect {
+        case Failure(regrets:RawlsExceptionWithErrorReport) => regrets.errorReport.causes
+      } flatten
+
+      // fail if nothing at all succeeded
+      if (numSuccesses == 0) {
+        // TODO: this could be a LOT of error reports, should we omit if large?
+        throw new RawlsExceptionWithErrorReport(RawlsErrorReport(StatusCodes.BadRequest, "All entities could not be updated.", failureReports))
```

Comment:
We used to have problems with big errors In Cromwell and found `take(100)` is a good balance between having something in the logs and not exploding

Reply:
hah! I was writing the `take(100)` implementation in the background and came back to see this comment ... perfect, thank you

Reply:
Haha nice!